x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/accesscontrol/FieldExtractor.java

```diff
@@ -35,68 +22,28 @@
-            }
-        } else if (query instanceof SpanTermQuery spanTermQuery) {
-            // we just do SpanTerm, other spans are trickier, they could contain
-            // the evil FieldMaskingSpanQuery: so SpanQuery.getField cannot be trusted.
-            fields.add(spanTermQuery.getField());
-        } else if (query instanceof TermQuery termQuery) {
-            fields.add(termQuery.getTerm().field());
-        } else if (query instanceof SynonymQuery q) {
-            // all terms must have the same field
-            fields.add(q.getTerms().get(0).field());
-        } else if (query instanceof PhraseQuery q) {
-            // all terms must have the same field
-            fields.add(q.getTerms()[0].field());
-        } else if (query instanceof MultiPhraseQuery q) {
-            // all terms must have the same field
-            fields.add(q.getTermArrays()[0][0].field());
-        } else if (query instanceof PointRangeQuery pointRangeQuery) {
-            fields.add(pointRangeQuery.getField());
-        } else if (query instanceof PointInSetQuery pointInSetQuery) {
-            fields.add(pointInSetQuery.getField());
-        } else if (query instanceof FieldExistsQuery fieldExistsQuery) {
-            fields.add(fieldExistsQuery.getField());
-        } else if (query instanceof IndexOrDocValuesQuery indexOrDocValuesQuery) {
-            // Both queries are supposed to be equivalent, so if any of them can be extracted, we are good
-            try {
-                Set<String> dvQueryFields = Sets.newHashSetWithExpectedSize(1);
-                extractFields(indexOrDocValuesQuery.getRandomAccessQuery(), dvQueryFields);
-                fields.addAll(dvQueryFields);
-            } catch (UnsupportedOperationException e) {
-                extractFields(indexOrDocValuesQuery.getIndexQuery(), fields);
-            }
-        } else if (query instanceof TermInSetQuery termInSetQuery) {
-            // TermInSetQuery#field is inaccessible
-            TermIterator termIterator = termInSetQuery.getTermData().iterator();
-            // there should only be one field
-            if (termIterator.next() != null) {
-                fields.add(termIterator.field());
+
+            @Override
+            public boolean acceptField(String field) {
+                fields.add(field);
+                return super.acceptField(field);
             }
-        } else if (query instanceof MatchAllDocsQuery) {
-            // no field
-        } else if (query instanceof MatchNoDocsQuery) {
-            // no field
-        } else {
-            throw new UnsupportedOperationException(); // we don't know how to get the fields from it
+        });
+        // Queries that don't touch any field and are not specifically known to do so are not supported
+        if (fields.isEmpty() && query instanceof MatchAllDocsQuery == false && query instanceof MatchNoDocsQuery == false) {
+            throw new UnsupportedOperationException();
         }
```

Comment:
I think this only works if the custom query is at the top level? What happens if it's nested in a boolean with another query that does report its fields?

Reply:
bummer :)

Reply:
ok I decided that it's not worth trying to cover for broken queries. A query can potentially have an empty visit method, and the visitor cannot do anything about that. I think that the only way is to trust that queries interact with visitors correctly, which is the case for all the queries we know of.